test/framework/src/main/java/org/elasticsearch/transport/nio/InboundChannelBuffer.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio;
+
+import org.elasticsearch.common.lease.Releasable;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.function.Supplier;
+
+public class InboundChannelBuffer {
+
+    public static final int PAGE_SIZE = 1 << 14;
+    private final int pageMask;
+    private final int pageShift;
+
+    private final ArrayDeque<Page> pages;
+    private final Supplier<Page> pageSupplier;
+
+    private int capacity = 0;
+    private int internalIndex = 0;
```

Comment:
should we rather use longs here since we can grow beyond an array an if not can we add safety guards to ensure it never does?